middleware/metrics_test.go

```diff
@@ -0,0 +1,138 @@
+package middleware
+
+import (
+	"io"
+	"net/http"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/gorilla/mux"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/prometheus/client_golang/prometheus/testutil"
+
+	"github.com/stretchr/testify/assert"
+)
+
+var tR = prometheus.NewCounterVec(
+	prometheus.CounterOpts{
+		Name: "http_requests_total",
+		Help: "Number of get requests.",
+	},
+	[]string{"path"},
+)
+
+var rS = prometheus.NewCounterVec(
+	prometheus.CounterOpts{
+		Name: "response_status",
+		Help: "Status of HTTP response",
+	},
+	[]string{"status"},
+)
+
+var hD = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+	Name: "http_response_time_seconds",
+	Help: "Duration of HTTP requests.",
+}, []string{"path"})
+
+func TestPrometheusRegisters(t *testing.T) {
+	assert := assert.New(t)
+
+	t.Log("Registering Prometheus Metrics")
+
+	prometheus.Register(tR)
+	prometheus.Register(responseStatus)
+	prometheus.Register(httpDuration)
+
+	tR.WithLabelValues("firstLabel").Inc()
+	tR.WithLabelValues("secondLabel").Inc()
+	tR.WithLabelValues("thirdLabel").Inc()
+	tR.WithLabelValues("thirdLabel").Inc()
+
+	rS.WithLabelValues("fourthlabel").Inc()
+	rS.WithLabelValues("fifthlabel").Inc()
+	rS.WithLabelValues("sixthlabel").Inc()
+
+	hD.WithLabelValues("seventhlabel").Observe(1)
+	hD.WithLabelValues("eighthlabel").Observe(2)
+	hD.WithLabelValues("ninelabel").Observe(2)
+
+	// tR collected three metrics
+	assert.Equal(3, testutil.CollectAndCount(tR))
+	// responseStatus collected three metrics
+	assert.Equal(3, testutil.CollectAndCount(rS))
+	// httpDuration collected three metrics
+	assert.Equal(3, testutil.CollectAndCount(hD))
+}
+
+func TestPrometheusMiddlewareAttached(t *testing.T) {
+	assert := assert.New(t)
+
+	t.Log("Launching server in go routine")
+	go func() {
+		t.Log("Defining router and registering prometheus targets")
+		router := mux.NewRouter()
+		router.Use(PrometheusMiddleware)
+		RegisterPrometheus()
+
+		router.Path("/metrics").Handler(promhttp.Handler())
+
+		t.Log("Launching server")
+		if err := http.ListenAndServe(":3030", router); err != nil {
```

Comment:
Does this ever exit? I feel like we would be leaking these http servers.

Reply:
The go routine closes when the outside function ends.

Reply:
This may not be as clean as I thought, actually. It definitely cleans itself up once `go test` exits, but it's not clear the server closes before then, which would make other tests... not good.

Reply:
Naked goroutines do not exit when they leave scope. Only when they stop running (or the process dies). An alternative here would be to use `httptest.NewRecorder()` like we do in `graphql_server\server_test.go`. Then you don't need this goroutine at all.

Reply:
Okay, tested it with this code block:
```go
package main

import (
	"fmt"
	"time"
)

func main() {
	go func() {
		count := 0
		for {
			fmt.Printf("Hello, playground, %d\n", count)
			count++
		}
	}()
	time.Sleep(5 * time.Nanosecond)
	fmt.Println("Going to dead time!")
}
```
It exits cleanly and doesn't keep running in the background.

Reply:
I got learned, and switched it to a httptest.Server, and use a defer close so we don't leave it running.